Termux (Android): `make test` failures with `/tmp`

```
make test
gcc -o testrun ./test/testrun.c
./es -s < ./test/test.es ./test/tests/*
file permissions: passed!
file types: passed!
matches match: passed!
redirects redirect: passed!
file globbing: passed!
match/if equivalence: passed!
subjects: passed!
error handling: passed!
es -c: passed!
es -e: passed!
es -p: passed!
es -i: passed!
syntactic sugar: passed!
heredoc sugar: passed!
match sugar: passed!
lexical analysis: test execution failure: error $&openfile /tmp/1665.lw: No such file or directory
^C
^C
^C
^C
^C
^C
^C
tokenizer errors
- {~ <={%backquote <={%flatten '' $ifs} {%dup 2 1 {$es -c 'echo hi |[2'}}} *^'expected ''='' or '']'' after digit'^*} failed signal sigint
- {~ <={%backquote <={%flatten '' $ifs} {%dup 2 1 {$es -c 'echo hi |[92='}}} *^'expected digit or '']'' after ''='''^*} failed signal sigint
- {~ <={%backquote <={%flatten '' $ifs} {%dup 2 1 {$es -c 'echo hi |[a]'}}} *^'expected digit after ''['''^*} failed signal sigint
- {~ <={%backquote <={%flatten '' $ifs} {%dup 2 1 {$es -c 'echo hi |[2-]'}}} *^'expected ''='' or '']'' after digit'^*} failed signal sigint
- {~ <={%backquote <={%flatten '' $ifs} {%dup 2 1 {$es -c 'echo hi |[2=99a]'}}} *^'expected '']'' after digit'^*} failed signal sigint
- {~ <={%backquote <={%flatten '' $ifs} {%dup 2 1 {$es -c 'echo hi |[2=a99]'}}} *^'expected digit or '']'' after ''='''^*} failed signal sigint
- {~ <={%backquote <={%flatten '' $ifs} {%dup 2 1 {$es -c 'echo ''hi'}}} *^'eof in quoted string'^*} failed signal sigint
- 0 cases passed, 7 failed.
blow the input stack: passed!
umask: test execution failure: error $&openfile /tmp/trip.1665: No such file or directory
redirections
- {~ <={%backquote <={%flatten '' \n} {%dup 2 1 {$es -c 'cat >[0=]'}}} *^'cat:'^*^'Bad file '^*} failed
- 6 cases passed, 1 failed.
exceptions: passed!
/tmp/big.1665: No such file or directory
cat: /tmp/big.1665: No such file or directory
heredocs and herestrings: test execution failure: error $&openfile /tmp/big.1665: No such file or directory
tilde matching: passed!
flat command expansion: passed!
equal sign in command arguments: passed!
exit status: passed!
wait is precise: passed!
setpgid: passed!
make: *** [Makefile:101: test] Error 1

```
```
mktemp
/data/data/com.termux/files/usr/tmp/tmp.s8gLzDqcuQ
```
Also, you have to send Ctrl-C several times for `make test` to continue

> 1. `mktemp` is clearly doing something wonky. I found [this link](https://github.com/termux/termux-packages/wiki/Common-porting-problems) which mentions "tmpnam() function may not respect TMPDIR variable and use hardcoded path to /tmp directory on some Android builds." That seems very suspicious to me. You should be able to resolve this by replacing any instances of `` `{mktemp blah.XXXXXX}`` with a random string such that the shell is working on files in the current directory.

I replaced instances of  `` `{mktemp blah.XXXXXX}`` with `./tmp/blah`,
and then it turned out there are hardcoded `/tmp` in `trip.es`, which I replaced with `./tmp`
https://github.com/Veraellyunjie/es-shell/tree/master/test/tests
```
mkdir -p tmp/test-glob-dir
make test
./es -s < ./test/test.es ./test/tests/*
file permissions: passed!
file types: passed!
matches match: passed!
redirects redirect: passed!
file globbing: passed!
match/if equivalence: passed!
subjects: passed!
error handling: passed!
es -c: passed!
es -e: passed!
es -p: passed!
es -i: passed!
syntactic sugar: passed!
heredoc sugar: passed!
match sugar: passed!
odd var formatting: passed!
lexical analysis: passed!
^C
^C
^C
^C
^C
^C
^C
tokenizer errors
- {~ <={%backquote <={%flatten '' $ifs} {%dup 2 1 {$es -c 'echo hi |[2'}}} *^'expected ''='' or '']'' after digit'^*} failed signal sigint
- {~ <={%backquote <={%flatten '' $ifs} {%dup 2 1 {$es -c 'echo hi |[92='}}} *^'expected digit or '']'' after ''='''^*} failed signal sigint
- {~ <={%backquote <={%flatten '' $ifs} {%dup 2 1 {$es -c 'echo hi |[a]'}}} *^'expected digit after ''['''^*} failed signal sigint
- {~ <={%backquote <={%flatten '' $ifs} {%dup 2 1 {$es -c 'echo hi |[2-]'}}} *^'expected ''='' or '']'' after digit'^*} failed signal sigint
- {~ <={%backquote <={%flatten '' $ifs} {%dup 2 1 {$es -c 'echo hi |[2=99a]'}}} *^'expected '']'' after digit'^*} failed signal sigint
- {~ <={%backquote <={%flatten '' $ifs} {%dup 2 1 {$es -c 'echo hi |[2=a99]'}}} *^'expected digit or '']'' after ''='''^*} failed signal sigint
- {~ <={%backquote <={%flatten '' $ifs} {%dup 2 1 {$es -c 'echo ''hi'}}} *^'eof in quoted string'^*} failed signal sigint
- 0 cases passed, 7 failed.
blow the input stack: passed!
umask
- umask 0 produces correct result failed
- umask 027 produces correct file failed
- 5 cases passed, 2 failed.
redirections: passed!
exceptions: passed!
heredocs and herestrings: passed!
tilde matching: passed!
flat command expansion: passed!
equal sign in command arguments: passed!
exit with signal codes: passed!
exit status: passed!
wait is precise: passed!
setpgid: passed!
make: *** [Makefile:101: test] Error 1

```

---

> 2. The `cat >[0=]` test failed. This is likely to be a trivial error message formatting issue. What is the output if you just run `cat >[0=]` from an *es* prompt?

```
; cat >[0=]
cat: -: Bad file descriptor
cat: closing standard input: Bad file descriptor
```

---

> 3. The hangs on the tokenizer errors. This is trickier; if it were my machine I'd want to jump in with `gdb`. As a start, a couple questions -- what happens if you try to run `echo hi |[2` from an *es* prompt? What happens if you try to run an `./es -c 'echo hi |[2'` from an *es* prompt?

```
echo hi |[2
expected '=' or ']' after digit
;
```
it prints `expected '=' or ']' after digit` and offers a new `;` prompt
```
./es -c 'echo hi |[2'
^C^C^C^C^C^C^C^C^C^C^C^C^C^C
```
it hangs, I sent Ctrl-C 20 times, nothing happened. Exited Termux via notification panel

I was able to scrounge up an Android device.  A few immediate notes:

1. `cat >&-` doesn't complain in bash either.  I don't actually think this one is anything specific to do with _es_.

2. There is a build warning (maybe you noted this somewhere already and I just missed it?)
```
token.c:81:24: warning: result of comparison of constant -1 with expression of type 'char' is always true [-Wtautological-constant-out-of-range-compare]
  81 |         while (c != '\n' && c != EOF)
     |
```
The implication of this is that we never break out of the `while` loop, and that is the cause of your problems with the "tokenizer errors" tests, given that line is in the function that gets called for tokenizer errors.  If you just replace the `char c` argument in token.c:80 with `int c`, it works.

3. The signals with pre-defined handlers are sigill, sigtrap, sigabrt, sigbus, sigfpe, sigsegv, sigstkflt, sigsys, sig35, sig36, and sig39.  (What are signals number 35, 36, and 39?  Hard for me to say.  Turns out most people who search for "android signal" are interested in other things.)  Based on what (at least most of) these signals have in common, it's likely that this handler is for coredump collection.  But that's not relevant to this issue -- what is relevant is that sigint is absent, which means my hunch about what was going wrong before was incorrect.  However, after I repro'd the "tokenizer error" bug and fix with `-funsigned-char` on my laptop, I noticed the fan was spinning up -- those seven _es_ processes from `make test` were still in the background, busy-looping away!

At this point the un-`^C`-ability issue clicked for me.

 - By default in Unix, `^C` kills a process immediately -- but in _es_ by default, `^C` is caught by the _es_ signal handler in order to be thrown as an exception later.

 - In order to actually throw the exception, _es_ needs to reach a `sigchk()` call, and when it's stuck in a tight loop like `token.c:81`, it never does.

 - When I asked you to delete `signal.c:202-203`, that removed the handle-and-throw behavior on `^C` for immediate-kill behavior, which is why it fixed the issue.

 - _Es_ only hits this infinite-loop bug on EOF, not on a newline, which is why a command at the interactive prompt (which is newline-terminated) doesn't cause any problems, but an `es -c` command, which is EOF-terminated, does.

 - So why are the buggy _es_ processes during `make test` `^C`-able?  That's because the test code is running them within a `` `{} `` (that is, the `$&backquote` primitive) and while a command is running within `$&backquote`, the parent _es_ is `read()`ing the child's output, not `wait()`ing for the child.  Why is that relevant?  Because in most cases, when a signal comes in during a `wait()`, the shell will go "yeah, okay, I'll log that away and throw an exception once this child exits".  But when a signal comes in during a `read()`, the shell will throw the exception right away!

That last behavior honestly seems like another bug (though it just happens to produce nicer behavior in this case).  Only `ksh` and `tcsh` share this behavior; `rc`, `dash`, `fish`, `zsh`, `bash`, and `elvish` all differ.

4. The `umask` test failure which I didn't even notice before:
```
umask
 - umask 0 produces correct result failed
 - umask 027 produces correct file failed
```
This seems to just be another trivial formatting bug.  `ls -l` on these files produces `-rw-rw-rw-.` and `-rw-r-----.`, respectively, whereas the test expects `-rw-rw-rw-` and `-rw-r-----`.

--- a/test/tests/trip.es
+++ b/test/tests/trip.es
@@ -77,13 +77,13 @@ test 'umask' {
 		umask 0
 		> $tmp
 		let (x = `{ls -l $tmp})
-			assert {~ $x(1) '-rw-rw-rw-'} umask 0 produces correct result
+			assert {~ $x(1) '-rw-rw-rw-'*} umask 0 produces correct result
 
 		rm -f $tmp
 		umask 027
 		> $tmp
 		let (y = `{ls -l $tmp})
-			assert {~ $y(1) '-rw-r-----'} umask 027 produces correct file
+			assert {~ $y(1) '-rw-r-----'*} umask 027 produces correct file
 
 		assert {~ `umask 027 0027} fail umask reports correct value
 	} {

--- a/token.c
+++ b/token.c
@@ -77,7 +77,7 @@ extern void print_prompt2(void) {
 }
 
 /* scanerror -- called for lexical errors */
-static void scanerror(char c, char *s) {
+static void scanerror(int c, char *s) {
 	while (c != '\n' && c != EOF)
 		c = GETC();
 	goterror = TRUE;

